modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/rule/FilterSpoolMergeToSortedIndexSpoolRule.java

```diff
@@ -75,13 +83,50 @@ private FilterSpoolMergeToSortedIndexSpoolRule(Config cfg) {
         if (F.isEmpty(idxCond.lowerCondition()) && F.isEmpty(idxCond.upperCondition()))
             return;
 
-        RelCollation collation = TraitUtils.createCollation(ImmutableList.copyOf(idxCond.keys()));
-        
+        RelCollation traitCollation;
+        RelCollation searchCollation;
+
+        if (inCollation == null || inCollation.isDefault()) {
+            // Create collation by index condition.
+            List<RexNode> lowerBound = idxCond.lowerBound();
+            List<RexNode> upperBound = idxCond.upperBound();
+
+            int cardinality = lowerBound != null ? lowerBound.size() : upperBound.size();
+
+            List<Integer> equalsFields = new ArrayList<>(cardinality);
+            List<Integer> otherFields = new ArrayList<>(cardinality);
+
```

Comment:
possibly we need to append smt like : 
`            if (lowerBound != null && upperBound != null)
                assert lowerBound.size() == upperBound.size();`

Reply:
Fixed